staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -0,0 +1,261 @@
+package filters
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	fctypesv1a1 "k8s.io/api/flowcontrol/v1alpha1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap"
+	"k8s.io/apiserver/pkg/authentication/user"
+	apifilters "k8s.io/apiserver/pkg/endpoints/filters"
+	epmetrics "k8s.io/apiserver/pkg/endpoints/metrics"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/server/mux"
+	utilflowcontrol "k8s.io/apiserver/pkg/util/flowcontrol"
+	fq "k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing"
+	fcmetrics "k8s.io/apiserver/pkg/util/flowcontrol/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+const (
+	decisionNoQueuingExecute = iota
+	decisionQueuingExecute
+	decisionCancelWait
+	decisionReject
+	decisionSkipFilter
+)
+
+type fakeApfFilter struct {
+	mockDecision int
+	postEnqueue  func()
+	postDequeue  func()
+}
+
+func (t fakeApfFilter) MaintainObservations(stopCh <-chan struct{}) {
+}
+
+func (t fakeApfFilter) Handle(ctx context.Context,
+	requestDigest utilflowcontrol.RequestDigest,
+	noteFn func(fs *fctypesv1a1.FlowSchema, pl *fctypesv1a1.PriorityLevelConfiguration),
+	queueNoteFn fq.QueueNoteFn,
+	execFn func(),
+) {
+	if t.mockDecision == decisionSkipFilter {
+		panic("Handle should not be invoked")
+	}
+	noteFn(bootstrap.SuggestedFlowSchemaGlobalDefault, bootstrap.SuggestedPriorityLevelConfigurationGlobalDefault)
+	switch t.mockDecision {
+	case decisionNoQueuingExecute:
+		execFn()
+	case decisionQueuingExecute:
+		queueNoteFn(true)
+		t.postEnqueue()
+		queueNoteFn(false)
+		t.postDequeue()
+		execFn()
+	case decisionCancelWait:
+		queueNoteFn(true)
+		t.postEnqueue()
+		queueNoteFn(false)
+		t.postDequeue()
+	case decisionReject:
+		return
+	}
+}
+
+func (t fakeApfFilter) Run(stopCh <-chan struct{}) error {
+	return nil
+}
+
+func (t fakeApfFilter) Install(c *mux.PathRecorderMux) {
+}
+
+func newApfServer(decision int, t *testing.T) *httptest.Server {
+	requestInfoFactory := &apirequest.RequestInfoFactory{APIPrefixes: sets.NewString("apis", "api"), GrouplessAPIPrefixes: sets.NewString("api")}
+	longRunningRequestCheck := BasicLongRunningRequestCheck(sets.NewString("watch"), sets.NewString("proxy"))
+
+	apfHandler := WithPriorityAndFairness(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if decision == decisionCancelWait {
+			t.Errorf("execute should not be invoked")
+		}
+		if decision != decisionSkipFilter && atomicReadOnlyExecuting != 1 {
+			t.Errorf("Wanted %d requests executing, got %d", 1, atomicReadOnlyExecuting)
+		}
+	}), longRunningRequestCheck, fakeApfFilter{
+		mockDecision: decision,
+		postEnqueue: func() {
+			if atomicReadOnlyWaiting != 1 {
+				t.Errorf("Wanted %d requests in queue, got %d", 1, atomicReadOnlyWaiting)
+			}
+		},
+		postDequeue: func() {
+			if atomicReadOnlyWaiting != 0 {
+				t.Errorf("Wanted %d requests in queue, got %d", 0, atomicReadOnlyWaiting)
+			}
```

Comment:
can you add a comment explaining that atomicReadOnlyWaiting can be either 1 or 0 because we are testing one request at a time.

Reply:
also, can we test the case where multiple requests waiting or executing?

Reply:
Added